app/services/single_logout_handler.rb

```diff
@@ -0,0 +1,76 @@
+require 'saml_idp/logout_request_builder'
+
+SingleLogoutHandler = Struct.new(:saml_response, :saml_request, :user) do
+  def successful_saml_response?
+    saml_response.present? && saml_response.success?
+  end
+
+  def finish_logout_at_idp?
+    failed_saml_response? || slo_not_implemented_at_sp?
+  end
+
+  def valid_saml_request?
+    saml_request.present? && saml_request.valid?
+  end
+
+  def request_message
+    Base64.strict_encode64(slo_request_builder(
+      sp_metadata,
+      user.uuid,
+      identity.session_uuid
+    ).signed)
+  end
+
+  def request_action_url
+    sp_metadata[:assertion_consumer_logout_service_url]
+  end
+
+  private
+
+  def failed_saml_response?
+    saml_response.present? && !saml_response.success?
+  end
+
+  def slo_not_implemented_at_sp?
+    identity.sp_metadata[:assertion_consumer_logout_service_url].nil?
+  end
+
+  def identity
+    return NullIdentity.new unless user
+
+    @_identity ||= begin
+      first_identity = user.first_identity
+
+      return first_identity if
```

Comment:
maybe Reek disagrees with me, but I find these `return foo if bar && baz` syntax constructs harder to maintain in the long run. I personally prefer
```
if saml_request && saml_request.issuer != first_identity.service_provider
  return
end
```

I know it's a bit longer but easier to drop a debug statement into later and easier to read, imo.